File: context-store/src/adapters/mongodb.ts
import type { ContextStore, SetOptions } from "../types.js";
import { withNamespace, stripNamespace, withTimeout } from "../config.js";

interface Options {
  url: string;
  namespace: string;
  collection: string;
  probeTimeoutMs: number;
}

interface MongoCollection {
  findOne(filter: Record<string, unknown>): Promise<{ value: string; expiresAt: number | null } | null>;
  updateOne(
    filter: Record<string, unknown>,
    update: Record<string, unknown>,
    opts: { upsert: boolean },
  ): Promise<unknown>;
  deleteOne(filter: Record<string, unknown>): Promise<{ deletedCount?: number }>;
  deleteMany(filter: Record<string, unknown>): Promise<unknown>;
  find(filter: Record<string, unknown>): { project(p: Record<string, 0 | 1>): { toArray(): Promise<Array<{ _id: string }>> } };
  createIndex(
    spec: Record<string, 1 | -1>,
    opts?: { partialFilterExpression?: Record<string, unknown> },
  ): Promise<unknown>;
}

interface MongoDb {
  collection(name: string): MongoCollection;
}

interface MongoClientLike {
  connect(): Promise<unknown>;
  close(): Promise<unknown>;
  db(): MongoDb;
}

interface MongoModule {
  MongoClient: new (url: string) => MongoClientLike;
}

async function loadMongo(): Promise<MongoModule> {
  try {
    return (await import("mongodb")) as unknown as MongoModule;
  } catch {
    throw new Error(
      "mongodb backend requested but the 'mongodb' package is not installed. " +
        "Run `npm install mongodb` or unset CCC_MONGO_URL.",
    );
  }
}

export async function openMongo({
  url,
  namespace,
  collection,
  probeTimeoutMs,
}: Options): Promise<ContextStore> {
  const mod = await loadMongo();
  const client = new mod.MongoClient(url);
  await withTimeout(client.connect(), probeTimeoutMs, "mongo connect");
  const col = client.db().collection(collection);
  await col.createIndex(
    { expiresAt: 1 },
    { partialFilterExpression: { expiresAt: { $type: "number" } } },
  );

  const sweep = async () => {
    await col.deleteMany({ expiresAt: { $lt: Date.now() } });
  };

  return {
    kind: "mongodb",
    async get(key) {
      await sweep();
      const doc = await col.findOne({ _id: withNamespace(namespace, key) });
      return doc ? doc.value : null;
    },
    async set(key, value, opts: SetOptions = {}) {
      const now = Date.now();
      const expiresAt = opts.ttlMs ? now + opts.ttlMs : null;
      await col.updateOne(
        { _id: withNamespace(namespace, key) },
        { $set: { value, expiresAt, updatedAt: now } },
        { upsert: true },
      );
    },
    async delete(key) {
      const result = await col.deleteOne({
        _id: withNamespace(namespace, key),
      });
      return (result.deletedCount ?? 0) > 0;
    },
    async list(prefix) {
      await sweep();
      const match = withNamespace(namespace, prefix ?? "");
      const filter = match ? { _id: { $regex: `^${escapeRegex(match)}` } } : {};
      const docs = await col.find(filter).project({ _id: 1 }).toArray();
      return docs.map((d) => stripNamespace(namespace, d._id));
    },
    async close() {
      await client.close();
    },
  };
}

function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

File: src/ccc/store/adapters/mongodb.py
"""MongoDB adapter (lazy import of `pymongo`)."""

from __future__ import annotations

import re
import time

from ccc.store.config import strip_namespace, with_namespace
from ccc.store.types import ContextStore


def _import_pymongo():
    try:
        import pymongo  # type: ignore
    except ImportError as exc:
        raise RuntimeError(
            "mongodb backend requested but the 'pymongo' package is not installed. "
            "Run `pip install pymongo` or unset CCC_MONGO_URL."
        ) from exc
    return pymongo


def _escape_regex(s: str) -> str:
    return re.escape(s)


class MongoAdapter(ContextStore):
    kind = "mongodb"

    def __init__(
        self,
        *,
        url: str,
        namespace: str,
        collection: str,
        probe_timeout_ms: int,
    ) -> None:
        pymongo = _import_pymongo()
        self._client = pymongo.MongoClient(
            url, serverSelectionTimeoutMS=probe_timeout_ms
        )
        # Force a server-selection round-trip so probes fail fast.
        self._client.admin.command("ping")
        db = self._client.get_default_database()
        if db is None:
            raise RuntimeError(
                "mongodb URL must include a database, e.g. mongodb://host:27017/dbname"
            )
        self._collection = db[collection]
        self._collection.create_index(
            "expiresAt",
            partialFilterExpression={"expiresAt": {"$type": "long"}},
        )
        self._namespace = namespace

    @staticmethod
    def _now_ms() -> int:
        return int(time.time() * 1000)

    def _sweep(self) -> None:
        self._collection.delete_many({"expiresAt": {"$lt": self._now_ms()}})

    def get(self, key: str) -> str | None:
        self._sweep()
        doc = self._collection.find_one({"_id": with_namespace(self._namespace, key)})
        return doc["value"] if doc else None

    def set(self, key: str, value: str, *, ttl_ms: int | None = None) -> None:
        now = self._now_ms()
        expires_at = now + ttl_ms if ttl_ms else None
        self._collection.update_one(
            {"_id": with_namespace(self._namespace, key)},
            {"$set": {"value": value, "expiresAt": expires_at, "updatedAt": now}},
            upsert=True,
        )

    def delete(self, key: str) -> bool:
        result = self._collection.delete_one(
            {"_id": with_namespace(self._namespace, key)}
        )
        return result.deleted_count > 0

    def list(self, prefix: str | None = None) -> list[str]:
        self._sweep()
        match = with_namespace(self._namespace, prefix or "")
        query = (
            {"_id": {"$regex": f"^{_escape_regex(match)}"}}
            if match
            else {}
        )
        return [
            strip_namespace(self._namespace, doc["_id"])
            for doc in self._collection.find(query, {"_id": 1})
        ]

    def close(self) -> None:
        try:
            self._client.close()
        except Exception:
            pass


def open_mongo(
    *, url: str, namespace: str, collection: str, probe_timeout_ms: int
) -> MongoAdapter:
    return MongoAdapter(
        url=url,
        namespace=namespace,
        collection=collection,
        probe_timeout_ms=probe_timeout_ms,
    )
